connaisseur/validators/cosign/cosign_validator.py

```diff
@@ -21,44 +22,110 @@
 class CosignValidator(ValidatorInterface):
     name: str
     trust_roots: list
+    vals: dict
     k8s_keychain: bool
 
     def __init__(self, name: str, trust_roots: list, auth: dict = None, **kwargs):
         super().__init__(name, **kwargs)
         self.trust_roots = trust_roots
+        self.vals = {}
         self.k8s_keychain = False if auth is None else auth.get("k8s_keychain", False)
 
-    def __get_key(self, key_name: str = None):
-        key_name = key_name or "default"
-        try:
-            key = next(
-                key["key"] for key in self.trust_roots if key["name"] == key_name
-            )
-        except StopIteration as err:
-            msg = 'Trust root "{key_name}" not configured for validator "{validator_name}".'
-            raise NotFoundException(
-                message=msg, key_name=key_name, validator_name=self.name
-            ) from err
-        return Key("".join(key))
-
     async def validate(
         self, image: Image, trust_root: str = None, **kwargs
     ):  # pylint: disable=arguments-differ
-        key = self.__get_key(trust_root)
-        return self.__get_cosign_validated_digests(image, key).pop()
 
-    def __get_cosign_validated_digests(self, image: Image, key: Key):
+        required = kwargs.get("required", [])
+        # if not configured, `threshold` is 1 if trust root is not "*" or
+        # `required` is specified and number of trust roots otherwise
+        threshold = kwargs.get(
+            "threshold",
+            (int(any(required)) or len(self.trust_roots)) if trust_root == "*" else 1,
+        )
+
+        self.vals = self.__get_pinned_keys(trust_root, required)
+
+        tasks = [self.__validation_task(k, image) for k in self.vals.keys()]
+        await asyncio.gather(*tasks)
+
+        return CosignValidator.__apply_policy(
+            vals=self.vals, threshold=threshold, required=required
+        )
+
+    def __get_pinned_keys(self, key_name: str = None, required: list = None):
+        """
+        Extract the pinned key(s) selected for validation from the list of trust roots.
+        """
+        key_name = key_name or "default"
+
+        if key_name == "*":
+            keys = {
+                k["name"]: {
+                    "name": k["name"],
+                    "key": Key("".join(k["key"])),
+                    "digest": None,
+                    "error": None,
+                }
+                for k in self.trust_roots
+            }
+
+            # check if all `required` are present in trust roots
+            not_found = set(required) - {k["name"] for k in self.trust_roots}
+            if not_found:
+                msg = 'Trust roots "{key_names}" not configured for validator "{validator_name}".'
+                raise NotFoundException(
+                    message=msg,
+                    key_names=", ".join(not_found),
+                    validator_name=self.name,
+                )
```

Comment:
This can go before the `keys` assignment above

Reply:
fixed as per comment and change below